File: core/FileIO.cpp
//--------------------------------------------------------------------//
// ファイル読み込み・書き込み
#include "FemDataModel.h"

#include <string>
#include <vector>
using namespace std;
using std::vector;



const string WHITESPACE = " \n\r\t\f\v";

string ltrim(const string& s)
{
    size_t start = s.find_first_not_of(WHITESPACE);
    return (start == string::npos) ? "" : s.substr(start);
}

string rtrim(const string& s)
{
    size_t end = s.find_last_not_of(WHITESPACE);
    return (end == string::npos) ? "" : s.substr(0, end + 1);
}

string trim(const string& s) {
    return rtrim(ltrim(s));
}


vector<string> split(string str, char del) {

    vector<string> result;
    stringstream ss{ str };
    string buf;
    while (getline(ss, buf, del)) {
        buf = trim(buf);
        if (buf.size() > 0) {
            result.push_back(buf);
        }
    }

    return result;
}


/// <summary>
/// // FEMデータを読み込む
/// </summary>
/// <param name="s">データ文字列のリスト</param>
/// <returns></returns>
FemDataModel readFemModel(string s)
{
    vector<string> lines = split(s, '\n');

    auto model = FemDataModel();
    model.clear();
    
    vector<Material> materials;
    vector<ShellParameter> shellParams;
    vector<BarParameter> barParams;
    vector<Coordinates> coordinates;
    vector<FENode> nodes;
    vector<ElementManager> elements;
    vector<Restraint> restraints;
    vector<Load> loads;
    vector<Pressure> pressures;
    vector<Temperature> temperature;
    vector<HeatTransferBound> htcs;


    vector<string> res;

    for (int i = 0; i < lines.size(); i++) {
        
        vector<string> columns = split(lines[i], ' ');

        if (columns.size() > 0) {

            string keyWord = columns[0];
            std::transform(keyWord.begin(), keyWord.end(), keyWord.begin(), ::tolower);

            // 材料データ
            if ((keyWord == "material") && (columns.size() > 7)) {
                auto mat = Material(stoi(columns[1]), stod(columns[2]),
                    stod(columns[3]), stod(columns[5]),
                    stod(columns[6]), stod(columns[7]));
                materials.push_back(mat);
            }
            /*
            // シェルパラメータ
            else if ((keyWord == "shellparameter") && (columns.size() > 2)) {
                shellParams.push
                (new ShellParameter(stoi(columns[1]), stod(columns[2])));
            }
            // 梁パラメータ
            else if ((keyWord == "barparameter") && (columns.size() > 4)) {
                barParams.push_back(new BarParameter
                (stoi(columns[1]), columns[2], ss.slice(3, columns.size())));
            }
            // 局所座標系
            else if ((keyWord == "coordinates") && (columns.size() > 10)) {
                coordinates.push_back(readCoordinates(ss));
            }
            */
            // 節点
            else if ((keyWord == "node") && (columns.size() > 4)) {
                auto node = FENode(stoi(columns[1]), stod(columns[2]),
                    stod(columns[3]),
                    stod(columns[4]));
                nodes.push_back(node);
            }
            // 要素
            /*
            else if ((keyWord == "bebarelement") && (columns.size() > 5)) {
                if (columns.size() < 8) {
                    elements.push_back(new BEBarElement
                    (stoi(columns[1]), stoi(columns[2]), stoi(columns[3]),
                        readVertex(ss, 4, 2)));
                }
                else {
                    elements.push_back(new BEBarElement
                    (stoi(columns[1]), stoi(columns[2]), stoi(columns[3]),
                        readVertex(ss, 4, 2),
                        new THREE.Vector3().set(stod(columns[6]),
                            stod(columns[7]),
                            stod(columns[8]))));
                }
            }
            else if ((keyWord == "tbarelement") && (columns.size() > 5)) {
                if (columns.size() < 8) {
                    elements.push_back(new TBarElement
                    (stoi(columns[1]), stoi(columns[2]), stoi(columns[3]),
                        readVertex(ss, 4, 2)));
                }
                else {
                    elements.push_back(new TBarElement
                    (stoi(columns[1]), stoi(columns[2]), stoi(columns[3]),
                        readVertex(ss, 4, 2),
                        new THREE.Vector3().set(stod(columns[6]),
                            stod(columns[7]),
                            stod(columns[8]))));
                }
            }
            else if ((keyWord == "trielement1") && (columns.size() > 6)) {
                elements.push_back(new TriElement1
                (stoi(columns[1]), stoi(columns[2]), stoi(columns[3]),
                    readVertex(ss, 4, 3)));
            }
            else if ((keyWord == "quadelement1") && (columns.size() > 7)) {
                elements.push_back(new QuadElement1
                (stoi(columns[1]), stoi(columns[2]), stoi(columns[3]),
                    readVertex(ss, 4, 4)));
            }
            else if ((keyWord == "tetraelement1") && (columns.size() > 6)) {
                elements.push_back(new TetraElement1
                (stoi(columns[1]), stoi(columns[2]), readVertex(ss, 3, 4)));
            }
            else if ((keyWord == "wedgeelement1") && (columns.size() > 8)) {
                elements.push_back(new WedgeElement1
                (stoi(columns[1]), stoi(columns[2]), readVertex(ss, 3, 6)));
            }
            */
            else if ((keyWord == "hexaelement1") && (columns.size() > 10)) {
                 auto elem = ElementManager("HexaElement1", columns);
                 elements.push_back(elem);
            }
            /*
            else if ((keyWord == "hexaelement1wt") && (columns.size() > 10)) {
                elements.push_back(new HexaElement1WT
                (stoi(columns[1]), stoi(columns[2]), readVertex(ss, 3, 8)));
            }
            else if ((keyWord == "tetraelement2") && (columns.size() > 12)) {
                elements.push_back(new TetraElement2
                (stoi(columns[1]), stoi(columns[2]), readVertex(ss, 3, 10)));
            }
            else if ((keyWord == "wedgeelement2") && (columns.size() > 17)) {
                elements.push_back(new WedgeElement2
                (stoi(columns[1]), stoi(columns[2]), readVertex(ss, 3, 15)));
            }
            else if ((keyWord == "hexaelement2") && (columns.size() > 22)) {
                elements.push_back(new HexaElement2
                (stoi(columns[1]), stoi(columns[2]), readVertex(ss, 3, 20)));
            }
            */
            // 境界条件
            else if ((keyWord == "restraint") && (columns.size() > 7)) {
                Restraint rest = Restraint(columns);
                restraints.push_back(rest);
            }
            else if ((keyWord == "load") && (columns.size() > 4)) {
                Load load = Load(columns);
                loads.push_back(load);
            }
            /*
            else if ((keyWord == "pressure") && (columns.size() > 3)) {
                pressures.push
                (new Pressure(stoi(columns[1]), columns[2].toUpperCase(),
                    stod(columns[3])));
            }
            else if ((keyWord == "temperature") && (columns.size() > 2)) {
                temperature.push
                (new Temperature(stoi(columns[1]), stod(columns[2])));
            }
            else if ((keyWord == "htc") && (columns.size() > 4)) {
                htcs.push
                (new HeatTransferBound(stoi(columns[1]), columns[2].toUpperCase(),
                    stod(columns[3]), stod(columns[4])));
            }
            // 計算結果
            else if ((keyWord == "resulttype") && (columns.size() > 1)) {
                if (columns[1].toLowerCase() == "element") {
                    model.result.type = ELEMENT_DATA;
                }
                else {
                    model.result.type = NODE_DATA;
                }
            }
            else if (((keyWord == "eigenvalue") && (columns.size() > 2)) ||
                ((keyWord == "displacement") && (columns.size() > 7)) ||
                ((keyWord == "strain1") && (columns.size() > 7)) ||
                ((keyWord == "stress1") && (columns.size() > 7)) ||
                ((keyWord == "strenergy1") && (columns.size() > 2)) ||
                ((keyWord == "strain2") && (columns.size() > 7)) ||
                ((keyWord == "stress2") && (columns.size() > 7)) ||
                ((keyWord == "strenergy2") && (columns.size() > 2)) ||
                ((keyWord == "temp") && (columns.size() > 2))) {
                res.push_back(ss);
            }
            */
        }
    }

    model.materials.resize(materials.size());
    model.shellParams.resize(shellParams.size());
    model.barParams.resize(barParams.size());
    model.coordinates.resize(coordinates.size());
    model.mesh.nodes.resize(nodes.size());
    model.mesh.elements.resize(elements.size());
    model.bc.restraints.resize(restraints.size());
    model.bc.loads.resize(loads.size());
    model.bc.pressures.resize(pressures.size());
    model.bc.temperature.resize(temperature.size());
    model.bc.htcs.resize(htcs.size());


    model.init();

    /*
    initObject();
    if (res.length > 0) {
        readFemResult(res);
    } else {
    */
    // model.calculate();
    // }
    

    return model;
}

File: tests/FileIO.py
from main import FEMPython

# テストデータを読み込む
def readTestData(s: str) -> FEMPython:

    fem = FEMPython()
    model = fem._model

    materials = []
    nodes = [] 
    elements = []
    shellParams = []
    barParams = []
    coordinates = []
    pressures = []
    temperature = []
    htcs = []
    restraints = []
    loads = []

    lines = s.splitlines()

    for l in lines:

        columns = l.split()

        columns_size = len(columns)
        if columns_size == 0:
            continue

        keyWord = columns[0].lower()

        # 材料データ
        if keyWord == "material" and columns_size > 7:
            mat = fem.core.Material(int(columns[1]), float(columns[2]),
                float(columns[3]), float(columns[5]),
                float(columns[6]), float(columns[7]))
            materials.append(mat)

        # シェルパラメータ
        elif keyWord == "shellparameter" and columns_size > 2:
            shell = fem.core.ShellParameter(int(columns[1]), float(columns[2]))
            shellParams.append(shell)

        # 梁パラメータ
        elif keyWord == "barparameter" and columns_size > 4:
            bar = fem.core.BarParameter(int(columns[1]), columns[2], columns[3:])
            barParams.append(bar)

        # 局所座標系
        elif keyWord == "coordinates" and columns_size > 10:
            cord = fem.core.readCoordinates(columns)
            coordinates.append(cord)

        # 節点
        elif keyWord == "node" and columns_size > 4:
            node = fem.core.FENode(str(columns[1]), float(columns[2]),
                float(columns[3]),
                float(columns[4]))
            nodes.append(node)
        
        # 要素
        elif keyWord == "bebarelement" and  columns_size > 5 \
          or keyWord == "tbarelement" and  columns_size > 5 \
          or keyWord == "trielement1" and  columns_size > 6 \
          or keyWord == "quadelement1" and  columns_size > 7 \
          or keyWord == "tetraelement1" and  columns_size > 6 \
          or keyWord == "wedgeelement1" and  columns_size > 8 \
          or keyWord == "hexaelement1" and  columns_size > 10 \
          or keyWord == "hexaelement1wt" and  columns_size > 10 \
          or keyWord == "tetraelement2" and  columns_size > 12 \
          or keyWord == "wedgeelement2" and  columns_size > 17 \
          or keyWord == "hexaelement2" and  columns_size > 22:
            elements.append(fem.core.ElementManager(keyWord, columns))
        
        # 境界条件
        elif keyWord == "restraint" and  columns_size > 7:
            rest = fem.core.Restraint(columns)
            restraints.append(rest)

        elif keyWord == "load" and  columns_size > 4:
            load = fem.core.Load(columns)
            loads.append(load)

        elif keyWord == "pressure" and  columns_size > 3:
            press = fem.core.Pressure(int(columns[1]), columns[2], float(columns[3]))
            pressures.append(press)

        elif keyWord == "temperature" and  columns_size > 2:
            temp = fem.core.Temperature(int(columns[1]), float(columns[2]))
            temperature.append(temp)

        elif keyWord == "htc" and  columns_size > 4:
            htc = fem.core.HeatTransferBound(int(columns[1]), columns[2],
                    float(columns[3]), float(columns[4]))
            htcs.append(htc)

        # 計算結果
        elif keyWord == "resulttype" and  columns_size > 1:
            keyWord1 = columns[1].lower()
            if keyWord1 == "element":
                model.result.type = model.result.ELEMENT_DATA
            else:
                model.result.type = model.result.NODE_DATA

    model.materials = materials
    # model.shellParams = shellParams
    # model.barParams = barParams
    # model.coordinates = coordinates

    model.mesh.nodes = nodes
    model.mesh.elements = elements

    model.bc.restraints = restraints
    model.bc.loads = loads
    #model.bc.pressures = pressures
    #model.bc.temperature = temperature
    #model.bc.htcs = htcs

    model.init()

    return fem